src/lib/support/BitMask.h

```diff
@@ -0,0 +1,134 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <assert.h>
+
+#include <limits>
+
+#include <lib/support/BitFlags.h>
+
+namespace chip {
+
+/**
+ * Stores bit masks and flags in a more typesafe manner.
+ *
+ * Extends BitFlags for boolean flags and also adds support for bit-mask set options.
+ *
+ */
+template <typename FlagsEnum, typename StorageType = typename std::underlying_type_t<FlagsEnum>>
+class BitMask : public BitFlags<FlagsEnum, StorageType>
+{
+public:
+    using IntegerType = typename BitFlags<FlagsEnum, StorageType>::IntegerType;
+
+    BitMask() : BitFlags<FlagsEnum, StorageType>() {}
+    BitMask(const BitFlags<FlagsEnum, StorageType> & other) : BitFlags<FlagsEnum, StorageType>(other) {}
+    BitMask(BitFlags<FlagsEnum, StorageType> && other) : BitFlags<FlagsEnum, StorageType>(std::move(other)) {}
+    BitMask(const BitMask &) = default;
+
+    explicit BitMask(FlagsEnum value) : BitFlags<FlagsEnum, StorageType>(value) {}
+    explicit BitMask(IntegerType value) : BitFlags<FlagsEnum, StorageType>(value) {}
+
+    template <typename... Args>
+    BitMask(FlagsEnum flag, Args &&... args) : BitFlags<FlagsEnum, StorageType>(flag, std::forward<Args>(args)...)
+    {}
+
+    template <typename... Args>
+    BitMask(IntegerType value, Args &&... args) : BitFlags<FlagsEnum, StorageType>(value, std::forward<Args>(args)...)
+    {}
+
+    BitMask & operator=(const BitMask &) = default;
+
+    BitMask & operator=(const BitFlags<FlagsEnum, StorageType> & other)
+    {
+        BitFlags<FlagsEnum, StorageType>::SetRaw(other.Raw());
+        return *this;
+    }
+
+    BitMask & operator=(FlagsEnum value)
+    {
+        BitFlags<FlagsEnum, StorageType>::SetRaw(static_cast<StorageType>(value));
+        return *this;
+    }
+
+    /**
+     * GetField to value via a mask shifting.
```

Comment:
This is documentation for SetField...  Also, I am not sure what "GetField to value via a mask shifting" means.

Reply:
Copy & paste got me. Updated all comments and added unit tests.